### DISCUSSION TITLE
OpenMP offload and HIP in the same translation unit. clang-15: error: unsupported option '-fopenmp-targets=' for language mode 'HIP'

### DISCUSSION THREAD

**Discussion by burlen** [ROCm/ROCm#2137] [Q&A] [answered]
I need to use OpenMP offload and HIP in the same translation unit. The two need to be compiled together due to use of templates. I've got a simple gnu make based reproducer ~100 lines of code, that lets one select between OpenMP only, HIP only, or OpenMP and HIP together, based on preprocessor defines. https://github.com/burlen/cuda_plus_offload

The case using OpenMP and HIP together fails to compile:
```
$ make -f Makefile.hip
...
hipcc -Wall -Wextra -g -DCUMP_USE_HIP -DCUMP_USE_OPENMP -D__HIP_PLATFORM_AMD__=1 -D__HIP_PLATFORM_HCC__=1 -isystem /opt/rocm-5.4.3/hip/include -x hip --offload-arch=gfx1030 -fopenmp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=gfx1030 main.cpp -o cump_both_hipcc
clang-15: error: unsupported option '-fopenmp-targets=' for language mode 'HIP'
make: *** [Makefile.hip:12: cump_both_hipcc] Error 1
```
Any help on the compiler options? Is it possible to use HIP and OpenMP in the same translation unit?

**Comment by ronlieb** [comment, marked as answer]
it is not currently possible to combine OpenMP offloading and HIP in the same translation unit.   This seemingly simple feature, requires an additional pass over the sourceTU to produce a 3rd device file and additional libraries. We do have an example of how to split the TU to get something similar to compile/link/run

see: https://github.com/ROCm-Developer-Tools/aomp/tree/aomp-dev/test/hip-openmp/matrixmul_omp_target_multifile

**Comment by burlen** [comment]
The approach shown in matrix multiplication example is probably not so useful when dealing with templates. In theory one could factor template declarations and implementations into different files and then explicitly instantiate every possible permutation of template argument types for every template class, template class method, and template function. But beyond the most simple function/class templates this really not practical. It also short circuits many of the benefits of templates.

At any rate, appreciate the fast answer, super helpful to know this isn't currently possible.
